### ISSUE TITLE
Upload is stay on 0 kb/s

### ISSUE THREAD

**Issue by burekas7** [elgatito/plugin.video.elementum#757]
Hi,

Why the upload rate is always on 0 kb/s even if I set a limit for example of 1000 kb/s.
The main idea of torrents is sharing.

![image](https://user-images.githubusercontent.com/13424257/119012605-dab1f880-b99e-11eb-8ac4-195c34c124a7.png)

Am I missing a setting or anything else?

Thanks.

---
Elementum 0.1.77
Kodi 17.6

**Comment by antonsoroko** [comment]
on your screenshot you can see that you have 0 connected peers. you connected only to seeds (who already has whole torrent).
peers don't have all torrent's data but probably you have some network issue like unpn is not working and double nat.
or simply they do not consider your host as "valuable" peer.

seeding it should be all the time.
but also it depends on your seeding limits (in settings).

**Comment by burekas7** [comment]
Thanks @antonsoroko 
I think this issue exist for a long time, but I will test it few time more and update.
I will check again the peers count and the upload rate if there are peers that connected.

But I have a question:
If I have for example 4/25 seeds, it means that I'm downloading from 4 users right.
But if it shows 3/20 peers, does it means that I'm downloading from 3 users and those 3 users are also downloading from me?

**Comment by burekas7** [comment]
@antonsoroko 
By the way, the setting of "Seed time limit (hours)" was on "0".
What does it mean? Does it relate to the setting "Seed forever" setting or in general?

**Comment by elgatito** [comment]
@burekas7 

> But if it shows 3/20 peers, does it means that I'm downloading from 3 users and those 3 users are also downloading from me?

You see connection numbers. 
Each connection is split into seeds/peers. 
Peers can upload to you pieces they have, can download from you, or do both.

**Comment by antonsoroko** [comment]
@burekas7 

after https://github.com/elgatito/elementum/pull/50 you will not need to set these 3 limits to `0` if you want to seed forever (only set "seed forever" itself). for now - you have to set all (3) limits to `0` or they will stop your torrents at some point (when limit is reached) even if you set "seed forever".

**Comment by antonsoroko** [comment]
@burekas7 also check your router upnp settings/table and elementum's bittorent settings tab.

probably you have some kind of misconfiguration and that's why you don't have any upload.

**Comment by burekas7** [comment]
@antonsoroko 

So let me summarize
Please confirm:

Seed Forever **(true)** : As long as the torrent exist in the list, seed all the time even after the download has been finished (Peers can still connect right?)
Seed Forever **(false)** : Seed only during the downloading progress, or seeding depends on the limit settings?

Limit settings:
Both "Seed Forever" options (true/false) depends on these limit options.

* I think you need to change the "Seed Forever" to "Seed Options" with multiple selection of:
- Forever (While Still activated in the torrents list)
- While Downloading
And add a comment above the limitation settings thay "0" means disable them.

---

* Regarding these limit settings: 
Regarding "Seed time limit (hours)":
For example setting time limit to 8 hours.
What does it mean? Each torrent starts a 8 hours timer since it started downloading?
(Currently set to 0)

What are the meaning of the other settings:
Seed  time ratio limit (percent) - Currently set to 200
Share ratio limit (percent) - Currently set to 25

---

And I have these settings in Elemntum enabled:
- Auto-assign network IP
- Auto-assign network ports
In my router upnp table there are some udp/tcp Int/Ext ports (all of them enabled 6883/4)
Should I need to add specific ports?

**Comment by antonsoroko** [comment]
@burekas7 

>Seed Forever (true) : As long as the torrent exist in the list, seed all the time even after the download has been finished (Peers can still connect right?)

yes.

>Seed Forever (false) : Seed only during the downloading progress, or seeding depends on the limit settings?

no. seeding depends on the limit settings.

>I think you need to change the "Seed Forever" to "Seed Options" with multiple selection of:

it is already changes. see PR above. so - wait for new release. if you will enable "seed forever" - no limits options shown (and obviously none are applied). otherwise limits are applied.

>Regarding these limit settings:

all 3 limits are described here
https://www.libtorrent.org/reference-Settings.html#seed_time_limit
https://www.libtorrent.org/reference-Settings.html#seed_time_limit

>Should I need to add specific ports?

no. if upnp table shows elementum ports (and ususally even a name of elementum) - then no.
but if you do not have "white" ip - it might be useless anyway.
you an read some materials in internet about how to check that you are ready to accept bittorent peers and what you can do about it.

**Comment by antonsoroko** [comment]
@burekas7 

>Seed Forever (true) : As long as the torrent exist in the list, seed all the time even after the download has been finished (Peers can still connect right?)

yes.

>Seed Forever (false) : Seed only during the downloading progress, or seeding depends on the limit settings?

no. seeding depends on the limits set in settings.

>I think you need to change the "Seed Forever" to "Seed Options" with multiple selection of:

it is already changed. see PR above. so - wait for new release. if you will enable "seed forever" - no limits options shown (and obviously none are applied). otherwise limits are applied.

>Regarding these limit settings:

all 3 limits are described here
https://www.libtorrent.org/reference-Settings.html#share_ratio_limit
https://www.libtorrent.org/reference-Settings.html#seed_time_ratio_limit
https://www.libtorrent.org/reference-Settings.html#seed_time_limit

just as a note: if i am not mistaken, libtorrent will account **real** seeding time (time that really was spent on seeding).
also, 0 means that limit disabled.

>Should I need to add specific ports?

no. if upnp table shows elementum ports (and ususally even a name of elementum) - then no.
but if you do not have "white" ip - it might be useless anyway.
you can read some materials in internet about how to check that you are ready to accept bittorent peers and what you can do about it.

**Comment by burekas7** [comment]
@antonsoroko 
> no. seeding depends on the limits set in settings.
But you said that "Seed Forever" also depends on the limitis (unless I change all of the limitiations to "0")

So both "Seed Forever" continue seeding except that when it set as "true" it should ignore the limitations and when it is "false" it should use them?

And I asked about the ports since in Elementum setting there are options for changing the ports manually.
But when using the "auto" option, what are the default ports?

But I will wait to the next version with the fixes.
Currently it seems to be ok and there is upload progress during the downloading.
When the download is finished the seeding/uploading stopped.

**Comment by antonsoroko** [comment]
>So both "Seed Forever" continue seeding except that when it set as "true" it should ignore the limitations and when it is "false" it should use them?

yes. at least for file storage. for memory - obviously it will stop seeding after you stop watching.

>But when using the "auto" option, what are the default ports?

you can see them in repo or locally 
https://github.com/elgatito/plugin.video.elementum/blob/27aed5a79b8b9e76fcadcb687d9e79dfea64529f/resources/settings.xml#L196-L197

>When the download is finished the seeding/uploading stopped.

right now - if you will set all limits to 0 it should not stop unless peers can't contact you or not interested in you as seeder.
after new version - just to enable  "Seed Forever" is enough.

**Comment by antonsoroko** [comment]
btw, i am not sure what you meant by (all of them enabled 6883/4)
port should be from 6891 to 6899.

if your port is 6883 then you probably also use quasar
https://github.com/scakemyer/plugin.video.quasar/blob/f8e1bc4deacad35d2807f3a31402f7e9160284cd/resources/settings.xml#L57-L58

ans this might be this issue.
